### DISCUSSION TITLE
Json, prisma, passthrough - since 3.21.0

### DISCUSSION THREAD

**Discussion by michalbundyra** [colinhacks/zod#2178] [Q&A] [answered]
Hello,

maybe this is my misuse, but since 3.21.0 I am having the following issue:

I am using prisma - JSON field:

```prisma
model MyModel {
  // ...
  json JSON
}
```

and then the following zod schema:

```typescript
const schema = z.object({
  name: z.string(),
}).passthrough()

const data = schema.parse({
  name: 'hello',
  extra: 'something',
})
```

and trying to insert it into the DB:

```typescript
ctx.prisma.my_model.create({
  data: {
    json: data,
  }
})
```

Then since 3.21.0 I am getting the following:

```
TS2322: Type 'objectOutputType<{ name: ZodString; }, ZodTypeAny, "passthrough">' is not assignable to type 'JsonNull | InputJsonValue'.   Type 'objectOutputType<{ name: ZodString; }, ZodTypeAny, "passthrough">' is missing the following properties from type 'readonly (InputJsonValue | null)[]': length, concat, join, slice, and 19 more.
```

Basically I want to store in the db whole json object, even if it contains some extra fields.
Maybe I was just misusing passthrough... Thanks

**Comment by JacobWeisenburger** [comment]
Everything is working fine for me. Perhaps this is a problem with Prisma types.

```ts
const schema = z.object( {
    name: z.string(),
} ).passthrough()
type Data = z.infer<typeof schema>
// type Data = {
//     name: string
// } & {
//     [ k: string ]: unknown
// }

console.log( schema.parse( { name: 'hello', extra: 'something' } ) )
// { name: 'hello', extra: 'something' }
```

If you found my answer satisfactory, please consider supporting me. Even a small amount is greatly appreciated. Thanks friend! 🙏
https://github.com/sponsors/JacobWeisenburger

**Comment by michalbundyra** [comment]
@JacobWeisenburger

thanks for your time and response.

I think `prisma` is right there, as prisma is expecting json object.
`zod` is also right here, as when we use passthrough we may have an additional unknown object, which may be not json-serialisable, so ... it's right as well.

I believe I should be using some kind of json validation instead of passthrough, as I my input is actually a json object.

Thanks!

**Comment by michalbundyra** [comment, marked as answer]
Just in case someone else looking for solution:

```typescript
// @see: https://github.com/colinhacks/zod#json-type
const literalSchema = z.union([z.string(), z.number(), z.boolean(), z.null()]);
type Literal = z.infer<typeof literalSchema>;
type Json = Literal | { [key: string]: Json } | Json[];
const jsonSchema: z.ZodType<Json> = z.lazy(() =>
  z.union([literalSchema, z.array(jsonSchema), z.record(jsonSchema)])
);

// and instead of passthrough use `.and(jsonSchema)`
const schema = z
  .object({
    name: z.string(),
  })
  .and(jsonSchema)
```

@JacobWeisenburger sorry, yeah, it should be moved to discussions.

Maybe worth adding it somewhere in the docs?
